clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -110,6 +110,19 @@ DeclBindingInfo *ResourceBindings::addDeclBindingInfo(const VarDecl *VD,
                                                       ResourceClass ResClass) {
   assert(getDeclBindingInfo(VD, ResClass) == nullptr &&
          "DeclBindingInfo already added");
+#ifndef NDEBUG
+  // Verify that existing bindings for this decl are stored sequentially
+  // and at the end of the BindingsList
+  auto I = DeclToBindingListIndex.find(VD);
+  if (I != DeclToBindingListIndex.end()) {
+    for (unsigned Index = I->getSecond(); Index < BindingsList.size(); ++Index)
+      assert(BindingsList[Index].Decl == VD);
+  }
+#endif
```

Comment:
I was thinking that `assert(!hasBindingInfoForDecl(VD) || BindingsList.back().Decl == VD);` would cover most bases?

Reply:
True

Reply:
Shoot! I have completed this PR before realizing I did not push this change up.
Here is a separate PR: https://github.com/llvm/llvm-project/pull/112661